### PR TITLE
add option to prevent waypoint extraction from note (fix #8078)

### LINKED ISSUE #8078: improve handling of automatic waypoint creation from notes
_(taken from: https://github.com/cgeo/cgeo/issues/8068#issuecomment-575601703 by @MagpieFourtyTwo)_

Nevertheless this "AutoZoom" (i. e. showing all defined waypoints of a cache) still renders the map view useless in some (but regrettably not really rare) cases, like e. g. in combination with c:geo's stubborn way of creating completely senseless waypoints from the personal note.

Example? Just create a personal note like this:

```
12-a-c=7	=> a+c=5	=> a=2 / c=3
36-b-d=26	=> b+d=10	=> b=2 / d=8

a.b.c.d@gmx.de
2.2.3.8@gmx.de (don't try, doesn't exist anyway ;) )

Final (doesn't exist, neither):
N48 51.234 E9 01.234
```
From this personal note two waypoints will be created automatically:
1. N48 51.234 E9 01.234 - which is perfectly fine
2. N02 12.000 E3 48.000 - which is *very* obviously *complete* rubbish

Usually you can suppress the automatic generating by putting a colon after the "N" like 
```
Calculation
N:48 51.234+17 E9 01.234-22
```
but in the example below even colons do not help ... :(

So, can we *please* invent a way of suppressing at least these completely senseless waypoints?
Perhaps it would be at least a quick'n dirty workaroud to suppress automatic generation when the extracted waypoint data gives a result which is more than let's say 10 km away from the base coordinates?  I guess, WPs more than 10 km away are rare, and if there is one, the user can create it using copy'n paste?

Which BTW at least for me would be the preferred way anyway, hence it could also be a solution to be able to switch this automatic extraction completely off by a config flag ...?

### PR DESCRIPTION
**Pull request by moving-bits**
(on per cache basis)

### CODE REVIEW

**Comment by bekuno** [comment]
Can you integrate it in edit note dialog?

**Comment by moving-bits** [comment]
> Can you integrate it in edit note dialog?

Which part would you like to be integrated in the edit note dialog?
The menu entry is intentionally put into the cache's menu, so that it is accessible from all cache details pages (especially the waypoint page).

**Comment by bekuno** [comment]
In this frame if possible:

**Comment by moving-bits** [comment]
@Bekuno:  What UI do you propose for this?

A third button or an overflow menu spring to my mind, but I'm not too happy with either of it:
- An overflow menu would sit at the right edge, where now the checkmark button is. - This might confuse some people which are used to confirm the note by pressing the upper right corner of the edit field. Also, this overflow menu would have only a single entry.
- A third button may be placed to the left of the two existing ones, but we would still end up with two different ways to access the same functionality, as IMHO we should have this functionality available at least on the waypoints page as well, where it would be (it is currently) a menu entry in the overflow menu.

(I haven't tested the technical aspects or checked available space yet.)

Maybe you have a good idea for how and where to place this item?

**Comment by bekuno** [comment]
@moving-bits 
We had also a discussion in #7215 for a automatic upload/sync of notes.
This can in my mind on this issue.

How about to add an option frame in this window above or below the text, and within a option button to enable/disable the extraction of coords.
We could also change the small dialog to a full size window like the LogCacheActivity.

**Comment by moving-bits** [comment]
> How about to add an option frame in this window above or below the text, and within a option button to enable/disable the extraction of coords.

I've extended the dialog box:

image

**Comment by bekuno** [comment]
Fine :)

As text I would say "prevent waypoint extraction"
"from note" not necessary, we are in the note dialog. 
".. extraction" - because we want waypoints in notes :)

**Comment by bekuno** [comment]
retest this please

File: main/src/cgeo/geocaching/models/Geocache.java [hunk 1]
```diff
@@ -1557,12 +1558,14 @@ public boolean addWaypointsFromNote() {
      * @param updateDb if true the added waypoints are stored in DB right away
      * @param namePrefix prefix for waypoint names
      */
-    public boolean addWaypointsFromText(@Nullable final String text, final boolean updateDb, @NonNull final String namePrefix) {
+    public boolean addWaypointsFromText(@Nullable final String text, final boolean updateDb, @NonNull final String namePrefix, final boolean forceExtraction) {
```
Comment (bekuno) [inline main/src/cgeo/geocaching/models/Geocache.java]: I do not understand of the parameter forceExtraction.
Why do we need them?

File: main/res/values/strings.xml [hunk 2]
```diff
@@ -857,6 +857,8 @@
     <string name="cache_log_images_title">Log images</string>
     <string name="cache_log_image_default_title">Photo</string>
     <string name="cache_personal_note">Personal note</string>
+    <string name="cache_menu_preventWaypointsFromNote">Prevent waypoints from note</string>
+    <string name="cache_menu_allowWaypointExtraction">Allow waypoints from note</string>
```
Comment (bekuno) [inline main/res/values/strings.xml]: Where is the extra parameter "cache_menu_allowWaypointExtraction" used?

Comment (moving-bits) [reply main/res/values/strings.xml]: Those two lines are for the cache menu - enabling / disabling waypoint extraction from personal note.

File: main/src/cgeo/geocaching/models/Geocache.java (hunk 1 above)
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Geocache.java]: `addWaypointsFromText()` is called in two contexts - extracting from listing (needs `forceExtraction = true`) and extracting from personal note (should respect global setting, therefore `forceExtraction = false`).

**Comment by moving-bits** [comment]
> As text I would say "prevent waypoint extraction"

You're right. In my changing the hard coded test-text to a resource I reused the text for the menu entry, which does not fit in this context. Changed the personal note edit label to this:

image

File: main/res/values/strings.xml (hunk 2 above)
Comment (bekuno) [reply main/res/values/strings.xml]: but it is a checkbutton, isn't it?

Comment (moving-bits) [reply main/res/values/strings.xml]: No, the strings starting with "cache_menu_" refer to the new menu entry at the end of the cache menu, whereas the new string "cache_personal_note_preventWaypointsExtraction" is for the checkbox in the personal note edit dialog.

Comment (bekuno) [reply main/res/values/strings.xml]: Understand, ok.

**Pull request merged**